fix_includes.py

```diff
@@ -1339,87 +1339,160 @@ def _GetFirstNamespaceLevelReorderSpan(file_lines):
 
-    # If we're a 'contentful' line such as a (non-header-guard) #ifdef, bail.
-    elif line_info.type in (_IF_RE, _NAMESPACE_END_RE, _ELSE_RE, _ENDIF_RE,
-                            None):    # None is a 'normal' contentful line
-      # TODO(csilvers): we could probably keep going if there are no
-      # braces on the line.  We could also keep track of our #ifdef
-      # depth instead of bailing on #else and #endif, and only accept
-      # the fwd-decl-inside-namespace if it's at ifdef-depth 0.
-      break
+    for line_number, line_info in enumerate(file_lines):
+      if line_info.deleted:
+        continue
 
-    elif line_info.type == _NAMESPACE_START_RE:
-      # Only handle the simple case of 'namespace <foo> {'
-      m = simple_namespace_re.match(line_info.line)
-      if not m:
-        break
-      namespace_prefix += ('namespace %s { ' % m.group(1).strip())
+      # If we're an empty line, just ignore us.  Likewise with #include
+      # lines, which aren't 'contentful' for our purposes, and the
+      # header guard, which is (by definition) the only kind of #ifdef
+      # that we can be inside and still considered at the "top level".
+      if line_info.type in (_COMMENT_LINE_RE,
+                            _BLANK_LINE_RE,
+                            _INCLUDE_RE,
+                            _HEADER_GUARD_RE,
+                            _HEADER_GUARD_DEFINE_RE,
+                            _PRAGMA_ONCE_LINE_RE):
+        continue
 
-    elif line_info.type in (_NAMESPACE_START_ALLMAN_RE,
-                            _NAMESPACE_START_MIXED_RE):
-      # Trim any possible comments on namespace to properly build
-      # the namespace level spans
-      m = simple_allman_namespace_re.match(line_info.line)
-      if not m:
-        break
-      namespace_prefix += ('namespace %s' % m.group(1).strip())
+      # If we're a 'contentful' line such as a (non-header-guard) #ifdef, add
+      # to the ifdef depth.  If we encounter #endif, reduce the ifdef depth.
+      # Only keep track of namespaces when ifdef depth is 0
+      elif line_info.type == _IF_RE:
+        ifdef_depth += 1
 
-    elif line_info.type == _NAMESPACE_CONTINUE_ALLMAN_MIXED_RE:
-      # Append to the simplified Allman namespace
-      namespace_prefix += ' { '
+      elif line_info.type == _ELSE_RE:
+        continue
 
-    elif line_info.type == _FORWARD_DECLARE_RE:
-      # If we're not in a namespace, keep going.  Otherwise, this is
-      # just the situation we're looking for!
-      if namespace_prefix:
-        return (namespace_prefix, line_info.reorder_span)
+      elif line_info.type == _ENDIF_RE:
+        ifdef_depth -= 1
+
+      elif ifdef_depth != 0:
+        continue # skip lines until we're outside of an ifdef block
+
+      # Build the simplified namespace lists and namespace depths.  When
+      # any new namespace is encountered, add the namespace to the list
+      # using the next line to cover namespaces without forward declares.
```

Comment:
Oh, actually, it looks like I did add documentation for this behavior.  Should I add a bit more detail?

Reply:
The documentation was actually a bit off as I changed patterns a bit after code review so I updated it.